nin/backend/generate/generate.js

```diff
@@ -1,21 +1,66 @@
 var fs = require('fs'),
     path = require('path'),
-    utils = require('../utils');
+    utils = require('../utils'),
+    mkdirp = require('mkdirp');
 
-var generate = function(toGenerate, path) {
-  var projectRoot = utils.findProjectRoot(path);
+var generate = function(type, name) {
+  var projectRoot = utils.findProjectRoot(process.cwd());
   if (projectRoot == '') {
     process.stderr.write('Could not find nin.json in project root\n');
     process.exit(1);
   }
 
-  if (endsWith(toGenerate, 'Layer')) {
-    generateLayer(toGenerate, projectRoot);
-    addToLayers(toGenerate, projectRoot);
+  if (name == "") {
+    return;
+  }
+
+  var camelizedName = utils.camelize(name);
+
+  switch (type) {
+    case 'layer':
+      generateLayer(camelizedName, 'TemplateLayer.js',
+          [[/TemplateLayer/g, camelizedName]],
+          projectRoot);
+
+      addToLayers(camelizedName, 'red', projectRoot);
+      break;
+
+    case 'shader':
+      generateShader(camelizedName, projectRoot);
+      break;
+
+    case 'shaderWithLayer':
+      var shaderLayerName = camelizedName + 'Layer';
+
+      generateShader(camelizedName, projectRoot);
+      generateLayer(shaderLayerName, 'TemplateShaderLayer.js',
+          [[/TemplateLayer/g, shaderLayerName],
+           [/TemplateShader/g, camelizedName]],
+          projectRoot);
+
+      addToLayers(camelizedName + 'Layer', 'pink', projectRoot);
```

Comment:
If you look in layers.json, we tend to not append Layer do the end of display name of layers only the type.

The displayname is used in the UI of nin, where we want as little clutter as possible.